File: apps/dashboard/src/lib/growth.ts
import type { RawMedia } from "./instagram";

// TS port of services/core/growth.py — turns raw IG media into growth signals.

export interface PostStat {
  id: string;
  caption: string;
  permalink: string;
  timestamp: string;
  likes: number;
  comments: number;
  reach: number;
  saves: number;
  views: number;
  productType: string;
  interactions: number;
  engagementRate: number;
}

export interface GrowthReport {
  postsAnalyzed: number;
  followers: number;
  mediaCount: number;
  avgEngagementRate: number;
  bestHourUtc: number | null;
  bestWeekday: string | null;
  bestFormat: string | null;
  topPosts: Array<{
    id: string;
    hook: string;
    permalink: string;
    engagementRate: number;
    reach: number;
  }>;
  recommendations: string[];
}

const WEEKDAYS = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];

export function toPostStat(m: RawMedia): PostStat {
  const likes = m.insights.likes ?? m.likeCount;
  const comments = m.insights.comments ?? m.commentsCount;
  const saves = m.insights.saved ?? 0;
  const reach = m.insights.reach ?? 0;
  const views = m.insights.views ?? 0;
  const interactions = likes + comments + saves;
  const engagementRate = reach > 0 ? Math.round((interactions / reach) * 10000) / 10000 : 0;
  return {
    id: m.id,
    caption: m.caption,
    permalink: m.permalink,
    timestamp: m.timestamp,
    likes,
    comments,
    reach,
    saves,
    views,
    productType: m.mediaProductType,
    interactions,
    engagementRate,
  };
}

function firstLine(caption: string): string {
  return (caption || "").trim().split("\n")[0]?.slice(0, 80) ?? "";
}

function bestBucket<T extends string | number>(buckets: Map<T, number[]>): T | null {
  let best: T | null = null;
  let bestAvg = -Infinity;
  for (const [key, vals] of buckets) {
    const avg = vals.reduce((a, b) => a + b, 0) / vals.length;
    if (avg > bestAvg) {
      bestAvg = avg;
      best = key;
    }
  }
  return best;
}

export function analyze(posts: PostStat[], followers: number, mediaCount: number): GrowthReport {
  if (posts.length === 0) {
    return {
      postsAnalyzed: 0,
      followers,
      mediaCount,
      avgEngagementRate: 0,
      bestHourUtc: null,
      bestWeekday: null,
      bestFormat: null,
      topPosts: [],
      recommendations: ["No posts to analyze yet."],
    };
  }

  const ranked = [...posts].sort((a, b) => b.engagementRate - a.engagementRate);
  const avgEr =
    Math.round((posts.reduce((s, p) => s + p.engagementRate, 0) / posts.length) * 10000) / 10000;

  const byHour = new Map<number, number[]>();
  const byWeekday = new Map<string, number[]>();
  const byFormat = new Map<string, number[]>();
  for (const p of posts) {
    const d = new Date(p.timestamp);
    if (!Number.isNaN(d.getTime())) {
      push(byHour, d.getUTCHours(), p.engagementRate);
      push(byWeekday, WEEKDAYS[d.getUTCDay()] ?? "Sun", p.engagementRate);
    }
    push(byFormat, p.productType, p.engagementRate);
  }

  const bestHour = bestBucket(byHour);
  const bestWeekday = bestBucket(byWeekday);
  const bestFormat = bestBucket(byFormat);

  const recommendations: string[] = [];
  if (bestHour !== null) {
    const ist = istLabel(bestHour);
    recommendations.push(
      `Post around ${pad(bestHour)}:00 UTC (${ist} IST) — your highest avg engagement window.`,
    );
  }
  if (bestWeekday) recommendations.push(`${bestWeekday} is your strongest day; prioritize key posts then.`);
  if (bestFormat) recommendations.push(`${bestFormat} outperforms other formats — lean into it.`);
  const hooks = ranked.slice(0, 3).map((p) => firstLine(p.caption)).filter(Boolean);
  if (hooks.length) recommendations.push(`Reuse winning hook patterns: ${hooks.map((h) => `"${h}"`).join("; ")}.`);
  if (avgEr < 0.02) recommendations.push("Engagement <2% — add a clear CTA and a stronger first line.");
  else if (avgEr > 0.06) recommendations.push("Strong engagement (>6%) — increase cadence to compound growth.");
  if (followers < 1000) recommendations.push("Under 1K followers: post 4–7×/week and engage in comments daily.");

  return {
    postsAnalyzed: posts.length,
    followers,
    mediaCount,
    avgEngagementRate: avgEr,
    bestHourUtc: bestHour,
    bestWeekday,
    bestFormat,
    topPosts: ranked.slice(0, 5).map((p) => ({
      id: p.id,
      hook: firstLine(p.caption),
      permalink: p.permalink,
      engagementRate: p.engagementRate,
      reach: p.reach,
    })),
    recommendations,
  };
}

function push<T>(map: Map<T, number[]>, key: T, val: number) {
  const arr = map.get(key) ?? [];
  arr.push(val);
  map.set(key, arr);
}
function pad(n: number): string {
  return String(n).padStart(2, "0");
}
function istLabel(utcHour: number): string {
  const totalMin = (utcHour * 60 + 5 * 60 + 30) % (24 * 60);
  const h = Math.floor(totalMin / 60);
  const m = totalMin % 60;
  return `${pad(h)}:${pad(m)}`;
}

File: services/core/growth.py
"""Turn raw post metrics into actionable growth recommendations.

Deterministic, cheap analysis (no LLM needed for the math). The learning agent
later uses these signals to tune prompts, posting times, and content cadence.
"""
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from datetime import datetime
from typing import Any


@dataclass
class PostStat:
    id: str
    caption: str
    timestamp: str  # ISO 8601 from the Graph API
    likes: int
    comments: int
    reach: int
    saves: int
    media_product_type: str
    permalink: str = ""

    @property
    def interactions(self) -> int:
        return self.likes + self.comments + self.saves

    @property
    def engagement_rate(self) -> float:
        if self.reach <= 0:
            return 0.0
        return round(self.interactions / self.reach, 4)

    @property
    def hour(self) -> int | None:
        try:
            return datetime.fromisoformat(self.timestamp.replace("Z", "+00:00")).hour
        except ValueError:
            return None

    @property
    def weekday(self) -> str | None:
        try:
            dt = datetime.fromisoformat(self.timestamp.replace("Z", "+00:00"))
            return ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"][dt.weekday()]
        except ValueError:
            return None


def _first_line(caption: str) -> str:
    return (caption or "").strip().splitlines()[0][:80] if caption else ""


def analyze(
    posts: list[PostStat],
    followers: int | None = None,
    media_count: int | None = None,
) -> dict[str, Any]:
    if not posts:
        return {"summary": "No posts to analyze yet.", "recommendations": []}

    ranked = sorted(posts, key=lambda p: p.engagement_rate, reverse=True)
    avg_er = round(sum(p.engagement_rate for p in posts) / len(posts), 4)

    # Best posting window by average engagement rate per hour bucket.
    by_hour: dict[int, list[float]] = defaultdict(list)
    by_weekday: dict[str, list[float]] = defaultdict(list)
    for p in posts:
        if p.hour is not None:
            by_hour[p.hour].append(p.engagement_rate)
        if p.weekday is not None:
            by_weekday[p.weekday].append(p.engagement_rate)

    best_hour = _best_bucket(by_hour)
    best_weekday = _best_bucket(by_weekday)

    # Format performance: reels vs feed.
    by_type: dict[str, list[float]] = defaultdict(list)
    for p in posts:
        by_type[p.media_product_type].append(p.engagement_rate)
    best_format = _best_bucket(by_type)

    recommendations = _recommend(
        avg_er=avg_er,
        best_hour=best_hour,
        best_weekday=best_weekday,
        best_format=best_format,
        top=ranked[:3],
        followers=followers,
    )

    return {
        "posts_analyzed": len(posts),
        "followers": followers,
        "media_count": media_count,
        "avg_engagement_rate": avg_er,
        "best_hour_utc": best_hour,
        "best_weekday": best_weekday,
        "best_format": best_format,
        "top_posts": [
            {
                "id": p.id,
                "hook": _first_line(p.caption),
                "permalink": p.permalink,
                "engagement_rate": p.engagement_rate,
                "reach": p.reach,
            }
            for p in ranked[:3]
        ],
        "recommendations": recommendations,
    }


def _best_bucket(buckets: dict[Any, list[float]]) -> Any | None:
    if not buckets:
        return None
    return max(buckets.items(), key=lambda kv: sum(kv[1]) / len(kv[1]))[0]


def _recommend(
    *,
    avg_er: float,
    best_hour: int | None,
    best_weekday: str | None,
    best_format: str | None,
    top: list[PostStat],
    followers: int | None,
) -> list[str]:
    recs: list[str] = []
    if best_hour is not None:
        recs.append(f"Post around {best_hour:02d}:00 UTC — your highest avg engagement window.")
    if best_weekday:
        recs.append(f"{best_weekday} is your strongest day; prioritize key posts then.")
    if best_format:
        recs.append(f"{best_format} outperforms other formats for you — lean into it.")
    if top:
        hooks = "; ".join(f'"{_first_line(p.caption)}"' for p in top if p.caption)
        if hooks:
            recs.append(f"Reuse winning hook patterns from top posts: {hooks}.")
    if avg_er < 0.02:
        recs.append("Engagement rate is low (<2%) — add a clear CTA and a stronger first line.")
    elif avg_er > 0.06:
        recs.append("Strong engagement (>6%) — increase cadence to compound growth.")
    if followers is not None and followers < 1000:
        recs.append(
            "Under 1K followers: post consistently (4-7x/week) and engage in comments daily."
        )
    return recs
